Translated chapter 11 ES to EN (#94)
btcven/locha: documentation/EN/chapter_11.md

- aodvv2_seqnum_init(void).
- aodvv2_seqnum_inc(void).
- aodvv2_seqnum_get().

Knowing the role that the sequence number plays in the _AODV_ protocol, we see the functions do no more than what their name says:

 - Starting a counter in **1** (init), as detailed in the protocol description.

 - Read the counter and get its value. (get)

 - Increase the counter by one (inc), as detailed in the protocol, for every _RREQ_ or _RREP_ that this node generates.

### 11.2.1 Routine to handle the sequence number.

```cpp
#include "net/aodvv2/seqnum.h"

#include <stdatomic.h>

static atomic_uint_fast32_t seqnum;

void aodvv2_seqnum_init(void)
{
    /* Initialize to 1 */
    atomic_init(&seqnum, 1);
}

void aodvv2_seqnum_inc(void)
{
    if (atomic_fetch_add(&seqnum, 1) >= 65535) {
        atomic_store(&seqnum, 1);
    }
}

aodvv2_seqnum_t aodvv2_seqnum_get(void)
{
    return atomic_load(&seqnum);
}

```

The library is very simple, but efficiently it solves possible data problems that can be detected when different execution threads in the application try to read and write the object that represents the sequence  number.

## 11.3 aodvv2_routing_table_init

> Todos los routers AODV deben mantener un conjunto de rutas locales, conteniendo información obtenida de los mensajes de ruta (RREQ y RREP).

Reviewer: This is in Spanish, it must be in English.

Author: updated, thanks 👍

> Todos los routers AODV deben mantener un conjunto de rutas locales, conteniendo información obtenida de los mensajes de ruta (RREQ y RREP).

Author: updated, thanks 👍

} aodvv2_local_route_t;
```
The ```CONFIG_AODVV2_MAX_ROUTING_ENTRIES``` directive represents the number of routes that can be maintained on the node. The resources available in the hardware have this limitation.

### 11.3.2 Routine for the initialization of the local routes table.
The routine in charge of starting the route table is defined in the file ```aodvv2_routingtable_init``` and the code is as follows::

```cpp
static aodvv2_local_route_t routing_table[CONFIG_AODVV2_MAX_ROUTING_ENTRIES];

static timex_t null_time;
static timex_t max_seqnum_lifetime;
static timex_t active_interval;
static timex_t max_idletime;
static timex_t validity_t;
static timex_t now;

void aodvv2_routingtable_init(void)
{
    DEBUG("aodvv2_routingtable_init()\n");

    null_time = timex_set(0, 0);
    max_seqnum_lifetime = timex_set(CONFIG_AODVV2_MAX_SEQNUM_LIFETIME, 0);
    active_interval = timex_set(CONFIG_AODVV2_ACTIVE_INTERVAL, 0);
    max_idletime = timex_set(CONFIG_AODVV2_MAX_IDLETIME, 0);
    validity_t = timex_set(CONFIG_AODVV2_ACTIVE_INTERVAL +
                           CONFIG_AODVV2_MAX_IDLETIME, 0);

    memset(&routing_table, 0, sizeof(routing_table));
}
```

> Starting the route table is equivalent to set to zero all the entries in the routes table and creating some time variables to measure if the information obtained from the routes message is obsolete or, on the contrary, provides us a useful information for the correct operation of the _AODV_ router.

Author: corregido 👍

For more information about routes table, check to section 4.5 of the ```draft-perkins-manet-aodvv2- document```.

## 11.4 aodv_rreqtable_init

> This function initializes a table with the intention of storing information from incoming _multicast_ messages to avoid retransmitting redundant messages. This table keeps a record of each incoming _RREQ_ message, so new incoming _RREQ_ messages, searching for the same destination from the same origin, are deleted without the need to retransmit them.

Author: corregido, thanks 👍

### 11.4.1 Scheme of an entry in the RREQ routing message table.
This is the format of the message to be stored in the routes message table (_RREQ_):

```cpp
typedef struct {
    struct netaddr origNode; /**< Node which originated the RREQ*/
    struct netaddr targNode; /**< Target (destination) of the RREQ */
    routing_metric_t metricType; /**< Metric type of the RREQ */
    uint8_t metric; /**< Metric of the RREQ */
    aodvv2_seqnum_t seqnum; /**< Sequence number of the RREQ */
    timex_t timestamp; /**< Last time this entry was updated */
} aodvv2_rreq_entry_t;

```

### 11.4.2 Routine for initializing the table of incoming Multicast messages.

This function is found in the file ```aodvv2_rreqtable.c```, which also exposes functions for the handling of said table.

> The code shows how some time variables are started and the zero value assignment to that table when the _AODV_ protocol starts.

Author: corregido 👍

The _mutex_ variable is used to block access to the table by other processes inside the application.

```cpp
static mutex_t rreqt_mutex;

static aodvv2_rreq_entry_t rreq_table[AODVV2_RREQ_BUF];

static timex_t _null_time;
static timex_t now;
static timex_t _max_idletime;

static struct netaddr_str nbuf;

void aodvv2_rreqtable_init(void)
{
    DEBUG("aodvv2_rreqtable_init()\n");
    mutex_lock(&rreqt_mutex);

    _null_time = timex_set(0, 0);
    _max_idletime = timex_set(CONFIG_AODVV2_MAX_IDLETIME, 0);
    memset(&rreq_table, 0, sizeof(rreq_table));
    mutex_unlock(&rreqt_mutex);
}
```

## 11.5 aodvv2_buffer_init

This function gives us the possibility of storing a packet inside a buffer  which there is no route for.

When a client wants to send a packet with a message the network stack first checks its routes table (NIB). If it can't find a route for that destination, you must start a _buffering process_ to store the packet inside the node until it finds the requested route.

> ### 11.5.1 Routine for packet buffer initialization.

Author: modificado 👍

The following code represents the function that is responsible for initializing the buffer where a message packet will be stored if a route to the destination is not known. The routine is inside the ```aodvv2_buffer.c``` file, which also offers routines for manipulating information.

```cpp
void aodvv2_buffer_init(void)
{
    memset(_buffered_pkts, 0, sizeof(_buffered_pkts));
}
```
### 11.5.2 Buffered message structure.

The message stored in the buffer must have the following structure:

```cpp
typedef struct {
    bool used;
    gnrc_pktsnip_t *pkt;
    ipv6_addr_t dst;
} buffered_pkt_t;
```

> ### 11.5.3 Definition of the buffer to store messages.

Author: modificado 👍

To configure messages that we want to sent but for which there is no route, an array is created with a fixed size and limited by hardware resources.

```
static buffered_pkt_t _buffered_pkts[CONFIG_AODVV2_MAX_BUFFERED_PACKETS];

```

> ## 11.6 Initiate the customers table.

Author: modificado 👍

> The clients table is a conceptual table that stores the clients of the router, and the node only recreates the _RREQ_ and _RREP_ messages from the clients registered in that table.

Author: 👍

        return -1;
    }
```

It could be the case that the node has more than one network interface. That is  why it is necessary to search for the global address in any of the interfaces available in the hardware.

### 11.8.1  _find_netif_global_addr.
For sending packets between nodes, you need to configure global IP addresses. If not, the network stack will not be able to route the packets correctly because all the **Turpial** boards will have two types of IPV6 addresses: a global and a local one.

```cpp
static int _find_netif_global_addr(ipv6_addr_t *addr)
{
    assert(addr != NULL);

    ipv6_addr_t addrs[CONFIG_GNRC_NETIF_IPV6_ADDRS_NUMOF];

    int numof = gnrc_netif_ipv6_addrs_get(_netif, addrs, sizeof(addrs));
    if (numof < 0) {
        DEBUG("aodvv2: couldn't get IPv6 addresses for iface\n");
        return -1;
    }

    for (unsigned i = 0; i < (numof / sizeof(ipv6_addr_t)); i++) {
        /* Pick up the first global address */
        if (ipv6_addr_is_global(&addrs[i])) {
            memcpy(addr, &addrs[i], sizeof(ipv6_addr_t));
            return 0;
        }
    }

    /* No address was found */
    return -1;
}
```

## 11.9 aodvv2_client_add.

> This function is used to add clients to the corresponding table. It is used for the first time inside the protocol initialization routine, due to the same device must be added as a client in this table.

Author: 👍

### 11.9.1 Routine to add clients to the table.

> This function receives as parameters the client's IP, the network prefix size and the cost of the link. The type of address passed to the function is of type ```ipv6_addr_t```. It stores 128-bit data. This clarification is made because _RIOT-OS_ offers other types of equivalent but not completely compatible addresses, such as ```netaddr```, which offers a little more information about the owner of the IP.

Author: 👍

> The following algorithm is simple, but it highlights the main task, which is to take the IP address of the argument in order to search in the clients table and know if it should update the information when there is a similar one, or must be created a new entry when not so.

Author: modificado 👍

          ipv6_addr_to_str(ipbuf, addr, sizeof(ipbuf)));

    aodvv2_client_entry_t *entry = aodvv2_client_find(addr);
    if (entry) {
        /* Update entry if it exists */
        mutex_lock(&_set_mutex);
        entry->prefix_length = prefix_length;
        entry->cost = cost;
        entry->_used = true;
        mutex_unlock(&_set_mutex);
        DEBUG("aodvv2_client_add_client: client is already stored\n");
        return entry;
    }

    /* Find free spot in client set and place the entry there */
    mutex_lock(&_set_mutex);
    for (unsigned i = 0; i < ARRAY_SIZE(_client_set); i++) {
        if (!_client_set[i]._used) {
            memcpy(&_client_set[i].ip_address, addr, sizeof(ipv6_addr_t));
            _client_set[i].prefix_length = prefix_length;
            _client_set[i].cost = cost;
            _client_set[i]._used = true;
            DEBUG("aodvv2_client_add_client: client added\n");
            mutex_unlock(&_set_mutex);
            return &_client_set[i];
        }
    }
    DEBUG("aodvv2_client_add_client: client table is full.\n");
    mutex_unlock(&_set_mutex);

    return NULL;
}
```

## 11.10 Registration in the network stack.

> In order to listen to incoming messages from external nodes, we must register to a specific type of messages, in this case _UDP_ messages. _RIOT-OS_ offers the possibility of having a thread from which you can listen to incoming messages using the _UDP_ protocol just by subscribing to such messages. This mechanism is known as _IPC_ (Inter Communication process), and can be configured to run synchronously or asynchronously.

Author: 👍

The difference between the type of configuration selected to listen to messages is that when we have messages received asynchronously, we enable a message queue which stores messages until they are read, respecting the size of the message queue.

In this fraction of code present in the initialization of the _AODV_ protocol, we are subscribing the node so that we can listen to all  _UDP_ messages from external nodes, without the need to configure _UDP_ on our own using the implementation that _RIOT-OS_ brings for handling of this type of subscriptions to external messages.

> In the function ```_gnrc_netreg_entry_init_pid```(& netreg, UDP_MANET_PORT, pid), pid refers to the thread ID that is designated for handling this type of message. As in the creation of any thread we need some standard parameters of the threads in C ++. To create the thread that is in charge of waiting for the messages, we proceed to the implementation by creating a parallel execution space with other processes inside _RIOT-OS_, known as _multi-task_.

Author: hecho 👍

### 11.10.1 Routine for subscribing to network messages.

> The term ```_event_loop``` in the next part of the code refers to the function that executes the code necessary to wait for incoming messages to which you have subscribed.

Author: 👍

If we see the following example, it is nothing different from what is proposed here for the register of network messages.

In this example, the main application is trying to send messages to the process that was created before.

```cpp
#include <inttypes.h>
#include <stdio.h>
#include "msg.h"
#include "thread.h"
#define RCV_QUEUE_SIZE  (8)
static kernel_pid_t rcv_pid;
static char rcv_stack[THREAD_STACKSIZE_DEFAULT + THREAD_EXTRA_STACKSIZE_PRINTF];
static msg_t rcv_queue[RCV_QUEUE_SIZE];
static void *rcv(void *arg)
{
    msg_t msg;
    (void)arg;
    msg_init_queue(rcv_queue, RCV_QUEUE_SIZE);
    while (1) {
        msg_receive(&msg);
        printf("Received %" PRIu32 "\n", msg.content.value);
    }
    return NULL;
}

int main(void)
{
    msg_t msg;
    msg.content.value = 0;
    rcv_pid = thread_create(rcv_stack, sizeof(rcv_stack),
                            THREAD_PRIORITY_MAIN - 1, 0, rcv, NULL, "rcv");
    while (1) {
        if (msg_try_send(&msg, rcv_pid) == 0) {
            printf("Receiver queue full.\n");
        }
        msg.content.value++;
    }
    return 0;
}

```

### 11.10.3 Implementation of the process or IPC thread.

> Now we will see how to implement the function that is responsible for listening the incoming messages, process them and review what kind of message has arrived and what should be the applied procedure, depending on the type and the information contained in it.

Author: corregido 👍

                DEBUG("AODVV2_MSG_TYPE_SEND_RREQ\n");
                {
                    aodvv2_msg_t m;
                    memcpy(&m, (aodvv2_msg_t *)msg.content.ptr, sizeof(m));
                    free(msg.content.ptr);

                    _send_rreq(&m.pkt, &m.next_hop);
                }
                break;

            case AODVV2_MSG_TYPE_SEND_RREP:
                DEBUG("AODVV2_MSG_TYPE_SEND_RREP\n");
                {
                    aodvv2_msg_t m;
                    memcpy(&m, (aodvv2_msg_t *)msg.content.ptr, sizeof(m));
                    free(msg.content.ptr);

                    _send_rrep(&m.pkt, &m.next_hop);
                }
                break;

            case GNRC_NETAPI_MSG_TYPE_RCV:
                DEBUG("GNRC_NETAPI_MSG_TYPE_RCV\n");
                _receive((gnrc_pktsnip_t *)msg.content.ptr);
                break;

            case GNRC_NETAPI_MSG_TYPE_GET:
            case GNRC_NETAPI_MSG_TYPE_SET:
                msg_reply(&msg, &reply);
                break;

            default:
                DEBUG("aodvv2: received unidentified message\n");
                break;
        }
    }

    /* Never reached */
    return NULL;
}

```

> The main structure of the function is formed by an infinite _while_ cycle, typical of the functions assigned to run in a separate thread or process, and a _switch_ that filters between the different incoming messages, to make decisions depending on the type of message.

Author: modificado 👍

From now on we will refer to this function for references to receiving _UDP_ messages, due to  this is where the internal process with incoming messages really starts.

```cpp
static void *_event_loop(void *arg)
{
    while(1) {
        msg_receive(&msg);

        switch (msg.type) {
            case AODVV2_MSG_TYPE_SEND_RREQ:
            break;
            case AODVV2_MSG_TYPE_SEND_RREP:
            break;
            case GNRC_NETAPI_MSG_TYPE_RCV:
            break;
            case GNRC_NETAPI_MSG_TYPE_GET:
            break;
            case GNRC_NETAPI_MSG_TYPE_SET:
            break;
            default:
            break;
        
        }
    }
}
``` 

At this point we have almost completely configured the resources necessary to operate on the **Locha Mesh** network. It´s only needed to configure the objects in charge of encoding and decoding the _AODV_ messages. These must travel through the network encapsulated in a format known as _RFC5444_, which is described further.

## 11.11 RFC5444 Initial Configuration.
 

> _RFC5444_ defines a message encapsulation and serialization protocol to operate in an ad-hoc mobile network, which provides the serialization and deserialization of messages and offers the possibility to read various parameters related to the packet without the need to deserialize  completely the packet.

Reviewer 2: ```suggestion
_RFC5444_ defines a message encapsulation and serialization protocol to operate in an ad-hoc mobile network, which provides  serialization and deserialization messages and offers the possibility to read various parameters related with the packet without the need completely packet  deserialization .
```

Author: modificado 👍

> _RFC5444_ was created to work with the _RPL_ protocol, which has been extended to different routing protocols, due to it adapts to the needs of this type of networks. It was designed for devices with low computational calculation and low energy consumption, which makes it ideal for this application.

Reviewer 2: ```suggestion
_RFC5444_ was created to work with the _RPL_ protocol, which has been extended to different routing protocols, due to it adapts to the needs of this type of networks. It was designed for low computational calculation and low energy consumption devices, which makes it ideal for this application.
```

Author: 👍
